### PR TITLE
Support interactive state edits

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Using `pulumi state {delete,rename,unprotect}` to perform state edits requires knowing the relevant resource URN ahead of time. It's common to know part of the URN you want (say, the name) but not have the full URN memorized. 

We should allow the user to omit the URN, and then allow the user to select from the list of valid URNs in the current stack.

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/cmd/pulumi/state.go [hunk 1]
```diff
@@ -193,3 +195,70 @@ func totalStateEdit(ctx context.Context, s backend.Stack, showPrompt bool, opts
...
+// stackName is the name of the current stack.
+//
+// snap is the snapshot of the current stack. snap cannot be nil, but (*snap) may be nil.
+// (*snap) will be set to the retrieved snapshot value is it is nil.
+//
+// Prompt is displayed to the user when selecting the URN.
+func getURNFromState(
+	ctx context.Context, stackName string, snap **deploy.Snapshot, prompt string,
+) (resource.URN, error) {
+	if snap == nil {
```
Comment (Frassle) [inline pkg/cmd/pulumi/state.go:210]: doc comment says "snap cannot be nil", but then your allowing it to be nil here. Suspect docs should be changed.

File: pkg/cmd/pulumi/state_rename.go [hunk 2]
```diff
@@ -128,32 +129,71 @@ func newStateRenameCommand() *cobra.Command {
 	var yes bool
 
 	cmd := &cobra.Command{
-		Use:   "rename <resource URN> <new name>",
+		Use:   "rename [resource URN] [new name]",
 		Short: "Renames a resource from a stack's state",
 		Long: `Renames a resource from a stack's state
 
 This command renames a resource from a stack's state. The resource is specified
-by its Pulumi URN (use ` + "`pulumi stack --show-urns`" + ` to get it) and the new name of the resource.
```
Comment (Frassle) [inline pkg/cmd/pulumi/state_rename.go:136]: I think we can leave a hint here about --show-urns but maybe rephrased to just comment that --show-urns can be used?

File: pkg/cmd/pulumi/state_rename.go [hunk 3]
```diff
@@ -128,32 +129,71 @@ func newStateRenameCommand() *cobra.Command {
...
+				}
+
+			case 2: // We got the URN and the name.
+				rName := args[1]
+				if !tokens.IsQName(rName) {
+					reason := "resource names may only contain alphanumerics, underscores, hyphens, dots, and slashes"
+					return result.Errorf("invalid name %q: %s", rName, reason)
+				}
+				newResourceName = tokens.QName(rName)
+			case 1: // We got the urn but not the name
```
Comment (Frassle) [inline pkg/cmd/pulumi/state_rename.go:177]: Don't we need a fallthrough here?

Comment (iwahbe) [reply pkg/cmd/pulumi/state_rename.go:177]: Not quite. Each case is actually independent. That said, this was a bug (now fixed).

File: pkg/cmd/pulumi/state_rename.go (hunk 2 above)
Comment (iwahbe) [reply pkg/cmd/pulumi/state_rename.go:136]: Sure

**Review by Frassle — APPROVED** [review]
